cosmwasm/ibc-union/lightclient/berachain/src/verify.rs

```diff
@@ -0,0 +1,152 @@
+    TimestampMustBeSet,
+    #[error("address must be set for commit sig with block id flag 4")]
+    AddressMustBeSet,
+    #[error("signature must exist for the committed validators")]
+    SignatureMustExist,
+    #[error("message must exist for all commits")]
+    MessageMustExist,
+    #[error("invalid public key type: {0}")]
+    InvalidPublicKeyType(String),
+    #[error("inner verification")]
+    InnerVerification(#[source] cosmwasm_std::VerificationError),
+    #[error("signature verification failed")]
+    SignatureVerificationFailed,
+    #[error(
+        "multiple signatures are provided, the light client is only capable of verifying a single aggregated signature"
+    )]
+    MultipleSignaturesAreProvided,
+    #[error("message must be set before doing the verification")]
+    MessageNotSet,
+    #[error("signature must be set before doing the verification")]
+    SignatureNotSet,
+}
+
+impl From<VerificationError> for Error {
+    fn from(value: VerificationError) -> Self {
+        Self::ClientSpecific(Box::new(value))
+    }
+}
+
+impl<'a> Verification for Bls12381Verifier<'a> {
+    type Error = VerificationError;
+
+    type CanonicalVoteProto = protos::cometbft::types::v1::CanonicalVote;
+
+    fn filter_commit(
+        &self,
+        commit_sig: &CommitSigRaw,
+    ) -> Result<Option<ValidatorSig>, Self::Error> {
+        if commit_sig.block_id_flag == 4 {
+            let Some(signature) = commit_sig.signature.clone() else {
+                return Ok(None);
+            };
+
+            Ok(Some(ValidatorSig {
+                validator_address: commit_sig
+                    .validator_address
+                    .ok_or(VerificationError::AddressMustBeSet)?
+                    .into_encoding(),
+                timestamp: Timestamp::default(),
+                signature: Some(signature.into_vec()),
+            }))
+        } else {
+            Ok(None)
+        }
+    }
+
+    fn process_signature(
+        &mut self,
+        public_key: PublicKey,
+        msg: Option<Vec<u8>>,
+        signature: Option<Vec<u8>>,
+    ) -> Result<(), Self::Error> {
+        // TODO(aeryz): verify here
+        let PublicKey::Bls12_381(public_key) = public_key else {
+            return Err(VerificationError::InvalidPublicKeyType(
+                "public key must be of type bls12381".to_string(),
+            ));
+        };
+        self.pubkeys.push(public_key.into_vec());
+
+        if msg.is_some() {
+            if self.msg.is_some() {
+                return Err(VerificationError::MultipleSignaturesAreProvided);
+            }
+            self.msg = msg;
+        }
+
+        if signature.is_some() {
+            if self.signature.is_some() {
+                return Err(VerificationError::MultipleSignaturesAreProvided);
+            }
+            self.signature = signature;
+        }
+
+        Ok(())
+    }
+
+    fn finish(&mut self) -> Result<(), Self::Error> {
```

Comment:
Since we had issues the last time in the wasm with specific host functions, i say we don't do unwraps here and name the errors properly to ease the debug if necessary.